Bug 1148835 - Remove toolbar in landscape

wesj opened the pull request:
This works pretty well. I ran down some rabbit holes so theres some work in here that probably isn't necessary (moving the Curve code to "widget" for instance).

thebnich commented on Utils/AppConstants.swift:
```diff
@@ -4,14 +4,14 @@
-    static let ToolbarHeight: CGFloat = 44
-    static let DefaultRowHeight: CGFloat = 58
 
+    static var ToolbarHeight: CGFloat = 44
```
Why change this from `let` to `var`?

wesj replied:
Heh. This number should be dynamic (and smaller in landscape). I changed it here once, but changed it back when things got hairy. Forgot the let->var change.

thebnich commented on Storage/SQL/JoinedHistoryVisitsTable.swift:
```diff
@@ -139,9 +139,6 @@ class JoinedHistoryVisitsTable: Table {
 
-        let dt2 = (NSDate().timeIntervalSince1970 - visit.date.timeIntervalSince1970) / 86400
-        println("DT \(site.url): \(dt2)")
-
```
Heh, this is definitely a tangent. It'd be helpful if you could split up these commits a little more in the future -- it's harder to read PRs with so many unrelated changes.

thebnich commented on Client/Frontend/Widgets/ToolbarTextField.swift:
```diff
@@ -0,0 +1,11 @@
+class ToolbarTextField: UITextField {
```
Missing license header.

thebnich commented on Client/Frontend/Widgets/InsetButton.swift:
```diff
@@ -0,0 +1,10 @@
+/**
+* Button whose insets are included in its intrinsic size.
+*/
```
Something looks out of place here.

thebnich commented on Client/Frontend/Widgets/InsetButton.swift:
```diff
@@ -0,0 +1,10 @@
+/**
```
Missing license header.

thebnich commented on Client/Frontend/Widgets/CurveView.swift:
```diff
@@ -0,0 +1,63 @@
+/* Code for drawing the urlbar curve */
```
Missing license header. You might want to consider making an Xcode template -- it always puts the Moz header at the top of new Swift files (aligned correctly!) so you don't have to worry about this anymore. I can share mine if you want.

wesj replied:
When I started to do this, I created a separate URLBarView implementation for landscape, and needed to reuse this on it. I pulled it out for reuse. (I still kinda wish I had those separate implementations, but the view/constraint lifecycle on iOS doesn't make it trivial to trade out views on rotation). I don't have super strong opinions, but I don't think the encapsulation granted here really gains us anything so I didn't put it back. i.e. we don't gain anything by saying "No one except the urlbar should ever use this widget for anything!" (as opposed to, "no one outside should know about or change this variable/method")

thebnich replied:
I'm not really worried about us breaking encapsulation in this case; I'd mostly just prefer it stay in `URLBarView` for organization reasons. It seems nicer to have one fewer class in the global namespace, and one fewer file in Widgets/. It just adds clutter if we put things in Widgets/ that, well...aren't widgets :)

To flip your "really gains us anything" statement around, the curve is already committed in the repo as a private class in `URLBarView`. Do we gain anything by introducing a change to pull it out?

thebnich commented on Client/Frontend/Browser/BrowserToolbar.swift:
```diff
@@ -6,28 +6,115 @@ import Foundation
 import Snap
 
+@objc
+protocol BrowserToolbarProtocol {
```
Why are you making this a protocol?

thebnich commented on Client/Frontend/Browser/BrowserToolbar.swift:
```diff
@@ -6,28 +6,115 @@ import Foundation
 
-private let ButtonInset = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
+@objc
+public class BrowserToolbarHelper {
```
What's the point of this helper? Between this and the protocol, it seems like you're trying to extract functionality here so we can reuse code between portrait/landscape modes. But I don't see you using either one in more than one place, so it's kind of confusing.

If you have plans to have multiple `BrowserToolbarProtocol` implementations that reuse the helper functionality, could you move these to that patch? Otherwise, what are these for?

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1024,6 +1074,7 @@ extension BrowserViewController: LongPressGestureDelegate {
         actionSheetController.addAction(cancelAction)
 
         if let popoverPresentationController = actionSheetController.popoverPresentationController {
+            println("Set sourceView")
```
Nit: remove

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -62,16 +61,51 @@ class BrowserViewController: UIViewController {
     }
 
+    private func shouldShowToolbar() -> Bool {
+        println("Should show \(traitCollection.verticalSizeClass.rawValue) != \(UIUserInterfaceSizeClass.Compact.rawValue) || \(traitCollection.horizontalSizeClass.rawValue) != \(UIUserInterfaceSizeClass.Regular.rawValue)")
```
Nit: remove

thebnich commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -115,28 +146,42 @@ class BrowserViewController: UIViewController {
+            make.top.equalTo(self.urlBar.snp_bottom)
+            make.left.right.equalTo(self.view)
+            let url = self.tabManager.selectedTab?.url
+            if url?.absoluteString == HomeURL {
```
Unfortunately, the condition isn't this straightforward. As the comment above says, "The home controller may change sizes if we tap the URL bar while on about:home.". That is, if we're on about:home, we show the bottom toolbar since we treat it like any other web page. But once we tap the URL bar, we're in "home mode", so the toolbar should disappear.

That was the reason the `#inline: Bool` argument was added to `showHomePanelController`. The location alone isn't enough information for us to make that decision.

thebnich commented on Client/Frontend/Browser/URLBarView.swift:
```diff
@@ -326,87 +426,17 @@ class URLBarView: UIView, BrowserLocationViewDelegate, UITextFieldDelegate {
-
-    override class func layerClass() -> AnyClass {
-        return CAShapeLayer.self
+extension URLBarView: BrowserToolbarProtocol {
```
It might make more sense to include this protocol at the `URLBarView` class definition in this case. All of the button getters are define there, so there's some disconnect between those and the methods down here (and I missed this extension on the first pass).

thebnich commented on Client/Frontend/Browser/URLBarView.swift:
```diff
@@ -37,28 +37,52 @@ class URLBarView: UIView, BrowserLocationViewDelegate, UITextFieldDelegate {
+        self.shareButton = UIButton()
+        self.bookmarkButton = UIButton()
+        self.forwardButton = UIButton()
+        self.backButton = UIButton()
```
Why are you duplicating all of these in each initializer? Can't you initialize all of them at the top level? That is:
```
let shareButton = UIButton()
let bookmarkButton = UIButton()
...
```

Also, can you define all of these buttons with `let` instead of `var`?

thebnich commented on Client/Frontend/Browser/URLBarView.swift:
```diff
@@ -120,10 +145,49 @@ class URLBarView: UIView, BrowserLocationViewDelegate, UITextFieldDelegate {
+                make.width.equalTo(AppConstants.ToolbarHeight)
+            }
+        } else {
+
```
Err...I assume there should be something here?

thebnich commented on Client/Frontend/Browser/URLBarView.swift:
```diff
@@ -37,28 +37,52 @@ class URLBarView: UIView, BrowserLocationViewDelegate, UITextFieldDelegate {
+    var forwardButton: UIButton
+    var backButton: UIButton
+    var helper: BrowserToolbarHelper?
+    var showToolbar = false
```
Nit: `showingToolbar`? `toolbarIsShown`?

`showToolbar` is an action, and sounds like a function.

thebnich commented on Client/Frontend/Browser/URLBarView.swift:
```diff
@@ -297,20 +376,41 @@ class URLBarView: UIView, BrowserLocationViewDelegate, UITextFieldDelegate {
+            self.forwardButton.hidden = !showToolbar || editing
+            self.backButton.hidden = !showToolbar || editing
+            self.shareButton.hidden = !showToolbar || editing
+            self.bookmarkButton.hidden = !showToolbar || editing
```
Ugh, this `if animated` is becoming really hard to read with these changes. One thing I noticed is that this block is identical to the `completion` block directly above it. It's also very similar to lines 357-362. Instead of duplicating all of this in 3 places, can you extract this to a helper method that takes a boolean and sets these accordingly?

wesj commented:
I moved the Curve code back into the urlbar file. TBH, I think in that case the encapsulation IS nice. i.e. we don't need to leak that info to everyone. But moving it out for me is actually more of a "There's too much crap in this URLBarView file" than anything.

I also split up the transitionToEditingMode() function to try and make it easier to read. Ideally, I'd like to take everywhere we have isToolbarShown checks and subclass urlbarView with them, but I spent a morning trying to switch correctly between the two views on rotation before I gave up.

For the share overlay, there seems to be an issue with telling the popover that it should point to the button. It works if I use the backButton, but nothing on the right side of the screen. I fixed it by actually anchoring it to the urlbar. That actually worked better (since otherwise, the arrow was pointing left/right instead of up).

I also had to create a member variable for the aboutHome state.
